### PR TITLE
Significant proof checking refactoring

### PR DESCRIPTION
**Pull request by K-dizzled**
The version increment is minor, yet the changes are significant. The main focus of this release was to improve interaction with the `coq-lsp` server. We now completely got rid of temporary files in the `extension` part of `CoqPilot`. This was done thanks to [ejgallego](https://github.com/ejgallego) and his improvements in `coq-lsp`. Now checking of proofs is done via the `proof/goals` request via `command` parameter, which is much more reliable and faster. Now a single `lsp` server is created for the plugin until being explicitly closed, and it tracks changes in the file. If you are using `coq-lsp` plugin itself, it will help you to always keep track of whether file is checked or not. When `CoqPilot` is ran on a completely checked file, it will not bring any significant overhead apart from requests to the Completion Services.

### CODE REVIEW

File: src/extension/coqPilot.ts [hunk 1]
```diff
@@ -30,44 +27,41 @@ import { Uri } from "../utils/uri";
...
-    constructor(vscodeExtensionContext: ExtensionContext) {
-        hideAuxFiles();
-        suggestAddingAuxFilesToGitignore();
-
-        this.vscodeExtensionContext = vscodeExtensionContext;
-        this.globalExtensionState = new GlobalExtensionState();
+    private readonly healthStatusIndicator: HealthStatusIndicator;
 
+    private constructor(
+        private readonly vscodeExtensionContext: ExtensionContext,
```
Comment (GlebSolovev) [inline src/extension/coqPilot.ts:61]: The contexts/states should be renamed for sure, now we have enough of them to get confused. For example:
* `vscodeExtensionContext: ExtensionContext` -> `vscodeContext: VSCodeContext`
* `globalExtensionState: GlobalExtensionState` -> `pluginState: PluginState`
* `sessionExtensionState: SessionExtensionState` -> `sessionState: SessionState`

Comment (GlebSolovev) [reply src/extension/coqPilot.ts:61]: Or maybe the global one should become `context`, if there is nothing that changes regularly

File: src/extension/coqPilot.ts [hunk 2]
```diff
@@ -81,70 +75,126 @@ export class CoqPilot {
...
-        this.performSpecificCompletionsWithProgress((_hole) => true, editor);
+        this.performSpecificCompletionsWithProgress(
+            (_hole) => true,
+            editor,
+            this.sessionExtensionState.abortController.signal
+        );
+    }
+
+    async toggleCurrentSession() {
+        if (this.globalExtensionState.hasActiveSession) {
```
Comment (GlebSolovev) [inline src/extension/coqPilot.ts:132]: Maybe it's better to move the activity status to the session state? Then the global state might work again as a context that only stores some data. Additionally, it's better to minimize dependencies between any components/interfaces

Comment (GlebSolovev) [reply src/extension/coqPilot.ts:132]: And yes, if we do so, then the `SessionState` itself will be responsible for setting its activity indicator to false after the `abort` / `dispose` call

Comment (GlebSolovev) [inline src/extension/coqPilot.ts:132]: I'm not sure if this `if` is enough: what will happen, if the user just _spams_ the `toggle` command?

File: src/extension/coqPilot.ts [hunk 3]
```diff
@@ -81,70 +75,126 @@ export class CoqPilot {
...
+                        ? EditorMessages.errorOccurred(e.message)
+                        : EditorMessages.objectWasThrownAsError(e),
+                    "error"
+                );
+                console.error(buildErrorCompleteLog(e));
             }
-        );
+        } finally {
+            this.healthStatusIndicator.hideSpinner(
+                this.globalExtensionState.hasActiveSession
```
Comment (GlebSolovev) [inline src/extension/coqPilot.ts:189]: Is this code truly reachable with the sessions being active?

File: src/extension/coqPilot.ts [hunk 4]
```diff
@@ -81,70 +75,126 @@ export class CoqPilot {
...
+                e.showAsMessageToUser();
+            } else if (e instanceof CoqLspStartupError) {
+                showMessageToUserWithSettingsHint(
+                    EditorMessages.coqLspStartupFailure(e.path),
+                    "error",
+                    `${pluginId}.coqLspServerPath`
+                );
+            } else if (
+                e instanceof CompletionAbortError ||
+                !this.globalExtensionState.hasActiveSession
```
Comment (GlebSolovev) [inline src/extension/coqPilot.ts:175]: Could you, please, explain what this check means here? Maybe in the code in the form of the comment also 👉👈

File: src/extension/coqPilot.ts [hunk 5]
```diff
@@ -314,7 +366,8 @@ export class CoqPilot {
     }
 
     dispose(): void {
-        cleanAuxFiles();
+        this.vscodeExtensionContext.subscriptions.forEach((d) => d.dispose());
         this.globalExtensionState.dispose();
+        this.sessionExtensionState.dispose();
```
Comment (GlebSolovev) [inline src/extension/coqPilot.ts:371]: Shouldn't it be disposed before the global one?

File: src/extension/editorMessages.ts [hunk 6]
```diff
@@ -31,6 +31,12 @@ export namespace EditorMessages {
     export const reportUnexpectedError = (errorDescription: string) =>
         `Coqpilot got an unexpected error: ${errorDescription}. Please report this crash by opening an issue in the Coqpilot GitHub repository.`;
 
+    export const completionAborted =
+        "Completion generation was forcefully aborted, Coq-LSP server stopping. Please hit the CoqPilot Status Bar button again to restart the server and proceed with completions.";
```
Comment (GlebSolovev) [inline src/extension/editorMessages.ts:35]: `Coq-LSP server stopping` -> `stopping Coq-LSP server` or `CoqLSP server is stopping`

Comment (GlebSolovev) [reply src/extension/editorMessages.ts:35]: Also, I feel like `proceed with completions` is a bit ambiguous here: will pressing the button just activate the CoqPlugin again or will it start performing the aborted request again?

File: src/extension/extensionAbortUtils.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+export class CompletionAbortError extends Error {
+    static readonly abortMessage =
+        "User has triggered a sesion abort: Stopping all completions";
```
Comment (GlebSolovev) [inline src/extension/extensionAbortUtils.ts:3]: `sesion` -> `session`

Comment (GlebSolovev) [reply src/extension/extensionAbortUtils.ts:3]: Also, shouldn't this message be `private`?

File: src/extension/globalExtensionState.ts [hunk 8]
```diff
@@ -14,8 +14,9 @@ import { EventLogger, Severity } from "../logging/eventLogger";
 import { pluginId } from "./coqPilot";
 import VSCodeLogWriter from "./vscodeLogWriter";
 
-export class GlobalExtensionState {
+export class GlobalExtensionState implements Disposable {
     public readonly eventLogger: EventLogger = new EventLogger();
+    public hasActiveSession = true;
```
Comment (GlebSolovev) [inline src/extension/globalExtensionState.ts:19]: Oh-oh, now it seems even more dangerous: a public modifiable property. A good reason to think about getting rid of it (by moving to another entity as `SessionState`, for example) or introducing a getter and a setter to make access more secure (by making it more explicit).

File: src/extension/healthStatusIndicator.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
+import {
+    ExtensionContext,
+    StatusBarAlignment,
+    StatusBarItem,
+    window,
+} from "vscode";
+
+import { pluginName } from "./coqPilot";
+
+export class HealthStatusIndicator {
```
Comment (GlebSolovev) [inline src/extension/healthStatusIndicator.ts:10]: Yep, for sure it's not the `Health` status: I feel it usually means that something can break down. Maybe our class should be more like `PluginStatusIndicator` or something like that

File: src/extension/healthStatusIndicator.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    private statusBarItem: StatusBarItem;
```
Comment (GlebSolovev) [inline src/extension/healthStatusIndicator.ts:11]: `readonly`

File: src/extension/healthStatusIndicator.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        if (isActive) {
+            this.statusBarItem.text = `$(debug-stop) ${pluginName}: Running`;
+            this.statusBarItem.tooltip = "Click to stop the extension";
+        } else {
+            this.statusBarItem.text = `$(debug-start) ${pluginName}: Stopped`;
+            this.statusBarItem.tooltip = "Click to start the extension";
+        }
+    }
+
+    showSpinner() {
```
Comment (GlebSolovev) [inline src/extension/healthStatusIndicator.ts:37]: `showInProgressSpinner`?

File: src/test/coqLsp/coqLspCommandInGoalsReq.test.ts
Comment (GlebSolovev) [inline src/test/coqLsp/coqLspCommandInGoalsReq.test.ts:1]: Maybe you could rename the file better? Unfortunately, I didn't get what you mean and what is the difference with the `coqLspGetGoals.test.ts`

File: src/test/coqLsp/coqLspCommandInGoalsReq.test.ts [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+    test("One Coq sentence: fail to solve with bad goal", async () => {
+        const goal = await getGoalAtPosition(
+            { line: 9, character: 4 },
+            ["small_document.v"],
+            "reflexivity."
+        );
+
+        expect(goal.err).toEqual(true);
```
Comment (GlebSolovev) [inline src/test/coqLsp/coqLspCommandInGoalsReq.test.ts:44]: The code that checks the goals extraction returned error should be moved in a separate method, so far there is pretty a lot of code duplication

File: src/test/coqLsp/coqLspCommandInGoalsReq.test.ts [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+    });
+
+    test("One Coq sentence wrapped into curly braces: solve successfully", async () => {
+        const goals = await getGoalAtPosition(
+            { line: 9, character: 4 },
+            ["small_document.v"],
+            " { auto. }"
+        );
+
+        expect(goals.ok).toEqual(true);
```
Comment (GlebSolovev) [inline src/test/coqLsp/coqLspCommandInGoalsReq.test.ts:122]: This block of code also repeats at least two times... Maybe you will even manage to move most of the successful checks into a single pretty abstract method 🙏

File: src/test/coqLsp/coqLspCommandInGoalsReq.test.ts [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+            "{ auto. }"
+        );
+
+        expect(goals.ok).toEqual(true);
+        if (goals.ok) {
+            expect(goals.val).toBeEmpty();
+        }
+    });
+
+    test("Many Coq sentences: good proof", async () => {
```
Comment (GlebSolovev) [inline src/test/coqLsp/coqLspCommandInGoalsReq.test.ts:156]: `good` -> `valid`, `bad` -> `invalid` / `incomplete`? (here and below)

File: src/test/core/completionGenerator.test.ts [hunk 15]
```diff
@@ -32,19 +32,31 @@ suite("Completion generation tests", () => {
             modelsParams: createPredefinedProofsModelsParams(predefinedProofs),
             services: createDefaultServices(),
         };
+        const abortController = new AbortController();
         try {
-            return await Promise.all(
+            await environment.coqLspClient.openTextDocument(
```
Comment (GlebSolovev) [inline src/test/core/completionGenerator.test.ts:37]: Shouldn't the code below be wrapped in the `try-finally` guard? (also relevant to the other `open`/`close` calls in the other tests)

**Review by GlebSolovev — CHANGES_REQUESTED** [review]
As always, great work with nice decisions on code structuring! 

Also, the whole love of all Coq-s 🐥, Pilot-s 🧑‍✈️ and especially me 🐲 for you, Andrei, for writing the tests for the new coq-lsp implementation ❤️

I left some comments on improving the code and making it a bit clearer, but in general everything is super solid and shiny 🌟

File: src/benchmark/framework/benchmarkingCore/singleCompletionGeneration/proofsCheckers/implementation/checkProofs.ts [hunk 16]
```diff
@@ -37,12 +38,20 @@ export namespace CheckProofsImpl {
...
-                args.sourceFileDirPath,
-                args.sourceFileContentPrefix,
-                args.prefixEndPosition,
+                fileUri,
+                args.documentVersion,
+                args.checkAtPosition,
                 args.preparedProofs
             );
+
+            await coqLspClient.closeTextDocument(fileUri);
```
[the remark this answers is not in the document]
Comment (K-dizzled) [reply src/benchmark/framework/benchmarkingCore/singleCompletionGeneration/proofsCheckers/implementation/checkProofs.ts:53]: I am not sure I understood the question( Well, in general, I guess the answer to your question is `yes`. The promised returned by method `checkProofs` could be rejected, yet, other errors shall not be thrown.

`closeDocument` ideally should happen in some wrapper over the `coqProofChecker`. Btw, would be neat to do smth like:
```typescript
withOpenAuxFile: 
    ...
```

File: src/benchmark/framework/parseDataset/coqProjectParser/implementation/parseCoqProject.ts [hunk 17]
```diff
@@ -74,13 +74,15 @@ export namespace ParseCoqProjectImpl {
         coqLspClient: CoqLspClient,
         logger: Logger
     ): Promise<SerializedParsedCoqFile> {
-        const mockFileVersion = 1;
+        const mockDocumentVersion = 1;
         const sourceFileUri = Uri.fromPath(filePath);
         await coqLspClient.openTextDocument(sourceFileUri);
+        const abortController = new AbortController();
```
[the remark this answers is not in the document]
Comment (K-dizzled) [reply src/benchmark/framework/parseDataset/coqProjectParser/implementation/parseCoqProject.ts:80]: I tried to do that but gave up on `executeOnParentProcessCall` as I didn't undertand the dependencies going on in the code. I left an `TODO` for you to do that 👉👈

File: src/coqParser/parseCoqFile.ts [hunk 18]
```diff
@@ -1,28 +1,45 @@
...
+/**
+ * TODO: [@Gleb Solovev] Refactor retrieveInitialGoal param
+ * to be something more reasonable and readable.
+ */
 export async function parseCoqFile(
     uri: Uri,
-    client: CoqLspClient
+    client: CoqLspClient,
+    abortSignal: AbortSignal,
+    retrieveInitialGoal: boolean = true
```
[the remark this answers is not in the document]
Comment (K-dizzled) [reply src/coqParser/parseCoqFile.ts:27]: I liked you chain of thought. I decided on `extractTheoremInitialGoal`, as I believe `extract` fits here better than `parse` or `save`

File: src/coqParser/parsedTypes.ts [hunk 19]
```diff
@@ -138,6 +138,7 @@ export class TheoremProof {
     }
 }
 
+// TODO: [LspCoreRefactor] Make Theorem.proof non nullable
```
[the remark this answers is not in the document]
Comment (K-dizzled) [reply src/coqParser/parsedTypes.ts:141]: I just was lazy. Did it now.

File: src/core/completionGenerator.ts [hunk 20]
```diff
@@ -195,29 +183,31 @@ async function checkAndFixProofs(
...
             prepareProofToCheck(generatedProof.proof())
     );
 
-    if (workspaceRootPath) {
-        processEnvironment.coqProofChecker.dispose();
-        const client = await createCoqLspClient(workspaceRootPath);
-        const coqProofChecker = new CoqProofChecker(client);
-        processEnvironment.coqProofChecker = coqProofChecker;
-    }
+    // TODO: Why is it happening every time?
```
[the remark this answers is not in the document]
Comment (K-dizzled) [reply src/core/completionGenerator.ts:196]: Well. The story was as following: I saw this piece of code, suspected that it's bad and useless, but remembered that when I wrote it it made sense. I inspected our code and still decided that its legacy but was not 100% sure. Anyways, deleted it, hopefully nothing will break.

File: src/extension/coqPilot.ts (hunk 2 above)
Comment (K-dizzled) [reply src/extension/coqPilot.ts:132]: Done. Everything until handling the updates of UI in SessionState was easy and neat. But accurately passing the StatusBar to the SessionState was kind of tricky in terms of dependencies. I hope its fine rn

Comment (K-dizzled) [reply src/extension/coqPilot.ts:132]: I don't want such users.............................

Comment (K-dizzled) [reply src/extension/coqPilot.ts:132]: What if user drops a nuke on his PC?

Comment (K-dizzled) [reply src/extension/coqPilot.ts:132]: You mean what happens if the client is still in the process of being created and the user switches the button on/off? Client boots quite fast, I don't see any sense in creating some new state like _bootingUp_

File: src/extension/coqPilot.ts (hunk 3 above)
Comment (K-dizzled) [reply src/extension/coqPilot.ts:189]: Yes, I checked that it works as expected

File: src/test/coqLsp/coqLspCommandInGoalsReq.test.ts
Comment (K-dizzled) [reply src/test/coqLsp/coqLspCommandInGoalsReq.test.ts:1]: Hard to do smth reasonable but I tried my best

**Pull request merged**